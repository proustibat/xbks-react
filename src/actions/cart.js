export const addBook = book => ( {
    type: 'ADD_BOOK',
    book
} );

export const removeBook = ( isbn = '' ) => ( {
    type: 'REMOVE_BOOK',
    isbn
} );

// Could be used if cart is saved in localstorage or cookie
export const setSavedBooks = books => ( {
    type: 'SET_SAVE_BOOKS',
    books
} );

export const changeQuantity = ( { isbn, qtt } ) => ( {
    type: 'CHANGE_QUANTITY',
    isbn,
    qtt
} );
