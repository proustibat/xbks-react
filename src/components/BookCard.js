import React from 'react';
import numeral from 'numeral';

export default class BookCard extends React.Component {
    constructor( props ) {
        super( props );

        this.state = {
            isbn: props.book ? props.book.isbn : '',
            title: props.book ? props.book.title : '',
            price: props.book ? ( props.book.price ).toString() : '',
            cover: props.book ? props.book.cover : '',
            synopsis: props.book ? props.book.synopsis : []
        }
    }

    render() {
        return (
            <div className="bookcard">
                <h1 className="bookcard__title">{ this.state.title }</h1>

                <div className="bookcard__row">
                    <div className="bookcard__cell-cover">
                        <img className="bookcard__cover" src={ this.state.cover } alt={ `cover for ${ this.state.title }` } />
                        <p className="bookcard__price">
                            <span>{ numeral( this.state.price ).format( '0.00$' ) }</span>
                        </p>
                    </div>
                    <div className="bookcard__cell-details">
                        <div className="bookcard__details">
                            <div className="bookcard__description">{ this.state.synopsis.join( ' ' ) }</div>
                        </div>
                    </div>
                </div>

            </div>
        );
    }
}

