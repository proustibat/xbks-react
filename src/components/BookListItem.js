import React from 'react';
import { Link } from 'react-router-dom';
import numeral from 'numeral';

const BookListItem = ( { isbn, title, price, synopsis } ) => (
    <Link to={ `/book/${ isbn }` } className="list-item">

        <div>
            <h3 className="list-item__title">{ title }</h3>
            <span className="list-item__sub-title">{ synopsis[ 0 ] }</span>
        </div>

        <h3 className="list-item__data">{ numeral( price ).format( '0.00$' ) }</h3>

    </Link>
);

export default BookListItem;
