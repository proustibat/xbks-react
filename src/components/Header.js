import React from 'react';
import Headroom from "react-headroom";
import { Link } from 'react-router-dom';

const Header = () => (
    <Headroom>
       <header className="header">
           <div className="content-container">
               <div className="header__content">
                   <Link className="header__title" to="/" >
                       <h1>XBKS</h1>
                   </Link>
                   <button className="button button--link">Cart</button>
               </div>
           </div>
       </header>
    </Headroom>
);

export default Header;
