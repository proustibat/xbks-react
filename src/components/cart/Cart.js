import React from 'react';
import CartList from './CartList';
import CartTotal from './CartTotal';

export class Cart extends React.Component {
    render () {
        return (
            <div className="cart">
                <CartList />
                <CartTotal />
            </div>
        );
    }
}

export default Cart;
