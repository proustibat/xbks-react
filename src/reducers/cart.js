const cartReducerDefaultState = {
    books: []
};

export default ( state = cartReducerDefaultState, action ) => {
    let books;
    switch ( action.type ) {
        case 'ADD_BOOK_TO_CART':
            const existingBook = state.books.find( book => book.isbn === action.book.isbn );
            if ( existingBook ) {
                existingBook.quantity = existingBook.quantity + 1;
                books = state.books;
            }
            else {
                action.book.quantity = 1;
                books = [ ...state.books, action.book ];
            }
            return { ...state, ...{ books } };
        case 'REMOVE_BOOK_FROM_CART':
            books = state.books.filter( ( { isbn } ) => action.isbn !== isbn );
            return { ...state, ...{ books } };
        // case 'SET_SAVED_BOOKS_TO_CART':
        //     action.books.forEach( book => {
        //         if ( !book.quantity ) {
        //             book.quantity = 1;
        //         }
        //     } );
        //     return { ...state, ...{ books: action.books } };
        case 'CHANGE_QUANTITY_BOOK_IN_CART':
            const booksList = [...state.books];
            let newBookList = booksList
                .map( book => {
                    book.quantity = book.isbn === action.isbn ? action.qtt : book.quantity;
                    return book;
                } )
                .filter( book => book.quantity > 0 );
            return { ...state, ...{ books: newBookList } };
        default:
            return state;
    }
};
