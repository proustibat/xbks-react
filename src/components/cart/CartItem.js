import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import { changeQuantity, removeBook } from '../../actions/cart';
import { startSetOffers } from "../../actions/offers";

export class CartItem extends React.Component {

    state = {
      subTotal: 0
    };

    constructor( props ) {
        super( props );
        this.state = {
            isbn: props.isbn || '',
            title: props.title || '',
            price: props.price || 0,
            quantity: props.quantity || 0
        }
    }

    handleNumberChange = ( e ) => {
        console.log( this.props.isbn, e.currentTarget.value );
        const quantity = parseInt( e.currentTarget.value, 10 );

        this.setState( () => ( { quantity } ) );

        this.props.changeQuantity( {
            isbn: this.state.isbn,
            qtt: quantity
        } );
        this.props.startSetOffers();
    };

    handleRemove = () => {
      this.props.removeBook( this.state.isbn );
      this.props.startSetOffers();
    };

    render() {
        return (
            <div className="cart-list-item">
            <span className="cart-list__col-item">
                <Link className="cart-list__col-item-title" to={ `/book/${ this.state.isbn }` }>{ this.state.title }</Link>
            </span>
            <span className="cart-list__col-qtt">
                <button className="cart-list__delete" onClick={ this.handleRemove }>Remove</button>
                <input type="number" min="0" id="itemsNumber" defaultValue={ this.state.quantity } step="1" onInput={ this.handleNumberChange } />
            </span>
                <span className="cart-list__col-unit-price">{ numeral( this.state.price ).format( '0.00$') }</span>
                <span className="cart-list__col-sub-total">{ numeral( this.state.quantity * this.state.price ).format( '0.00$' ) }</span>
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => ( {
    changeQuantity: ( { isbn, qtt } ) => dispatch( changeQuantity( { isbn, qtt } ) ),
    removeBook: isbn => dispatch( removeBook( isbn ) ),
    startSetOffers: () => dispatch ( startSetOffers() )
} );

export default connect( undefined, mapDispatchToProps )( CartItem );
