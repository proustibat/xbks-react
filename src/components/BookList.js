import React from 'react';
import { connect } from 'react-redux';
import BookListItem from './BookListItem';
import getVisibleBooks from "../selectors/getVisibleBooks";

export class BookList extends React.Component {
    render() {
        return (
            <div className="booklist content-container">
                { this.props.books.length === 0 ? (
                    <span>No Books</span>
                ) : (
                    this.props.books.map( book => (
                        <BookListItem
                            key = { book.isbn }
                            { ...book }
                        />
                    ) )
                ) }
            </div>
        );
    }
}

const mapStateToProps = state => ( {
    books: getVisibleBooks( state.books, { searchTerm: state.filters.searchTerm } )
} );

export default connect( mapStateToProps )( BookList );
