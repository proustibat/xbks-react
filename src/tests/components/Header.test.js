import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';
import cart from "../fixtures/cart";

test( 'Should render Header component correctly with items in cart', () => {
    const wrapper = shallow( <Header items={ cart.books.length } /> );
    expect( wrapper ).toMatchSnapshot();
} );

test( 'Should render Header component correctly with no any items in cart', () => {
    const wrapper = shallow( <Header items={ [] } /> );
    expect( wrapper ).toMatchSnapshot();
} );
