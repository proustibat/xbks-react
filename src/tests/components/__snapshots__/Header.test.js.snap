// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Header component correctly with items in cart 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/"
      >
        <h1>
          XBKS
        </h1>
      </Link>
      <Link
        className="button button--link"
        replace={false}
        to="/cart"
      >
        Cart 
        <span
          className="header__cart-info--items"
        >
          <br />
          2
           items
        </span>
      </Link>
    </div>
  </div>
</header>
`;

exports[`Should render Header component correctly with no any items in cart 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/"
      >
        <h1>
          XBKS
        </h1>
      </Link>
      <Link
        className="button button--link"
        replace={false}
        to="/cart"
      >
        Cart 
      </Link>
    </div>
  </div>
</header>
`;
